### PR TITLE
Heartbeat interval handled by library instead of vehicle

### PR DESCRIPTION
**Pull request by magicrub**
Heartbeat interval handled by library instead of vehicle.

There's one additional special handling which could be removed in [mavlink_delay_cb](https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/GCS_Mavlink.cpp#L1279).

### CODE REVIEW

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 1]
```diff
@@ -173,6 +173,8 @@ GCS_MAVLINK::setup_uart(const AP_SerialManager& serial_manager, AP_SerialManager
         // after experiments with MAVLink2
         status->flags |= MAVLINK_STATUS_FLAG_OUT_MAVLINK1;
     }
+
+    set_mavlink_message_id_interval(MAVLINK_MSG_ID_HEARTBEAT, 1000);
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: This shouldn't be here; under the call to `initialise_message_intervals_from_streamrates()` would be more appropriate.

File: libraries/GCS_MAVLink/GCS.h [hunk 2]
```diff
@@ -287,6 +287,12 @@ class GCS_MAVLINK
 
     virtual uint64_t capabilities() const;
 
+    // set an inverval for a specific mavlink message.  Returns false
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS.h]: This change (making this method public) doesn't seem necessary.  It's not necessarily a bad thing, but I don't think we need it for this PR, and the longer we keep stuff non-public the easier to update/fix things.

Comment (magicrub) [reply libraries/GCS_MAVLink/GCS.h]: hmm... looked like it was. My bad. I'll revert that part

**Comment by magicrub** [comment]
0 Hz allows 0 hertz, that's bad so I'll fix that while I'm at here
10Hz sends at 10Hz. 
>=40Hz or -1 sends at 40Hz.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 1 above)
Comment (magicrub) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: fixed

File: libraries/GCS_MAVLink/GCS.h (hunk 2 above)
Comment (magicrub) [reply libraries/GCS_MAVLink/GCS.h]: fixed

**Comment by magicrub** [comment]
I think the only thing remaining on this, is if we want to give the user the option to be slower than 1Hz. Faster is OK, GCSs seems happy with that because we run the simulators faster than realtime, but slower will cause all sorts of timeouts to start triggering. So, shall we restrict to >= 1000ms? What about sat links that would be happier with less data?

**Comment by peterbarker** [comment]
On Fri, 3 May 2019, Tom Pittenger wrote:

> I think the only thing remaining on this, is if we want to give the user the option to be slower than 1Hz. Faster is OK, GCSs seems happy with that
> because we run the simulators faster than realtime, but slower will cause all sorts of timeouts to start triggering. So, shall we restrict to >= 1000ms?
> What about sat links that would be happier with less data?

Yeah, I don't think we should add *too* much magic :-)

**Comment by magicrub** [comment]
@peterbarker I've addressed your questions. Anything else blocking this?

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 3]
```diff
@@ -2298,7 +2298,12 @@ MAV_RESULT GCS_MAVLINK::handle_command_set_message_interval(const mavlink_comman
     const int32_t interval_us = (int32_t)packet.param2;
 
     uint16_t interval_ms;
-    if (interval_us == 0) {
+
+    if (msg_id == MAVLINK_MSG_ID_HEARTBEAT && (interval_us <= 1000000)) {
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: This structure means you can never reset the heartbeat to its default rate.  Also, some links might really, really not want heartbeat and we should allow that.  I'm thinking sattelite high-latency-high-cost.

**Comment by peterbarker** [comment]
@magicrub you do need sign-off from the vehicle maintainers here.  FWIW I'm happy with this for tracker :-)

**Comment by magicrub** [comment]
we've been using this on our fork. works well so far. I was actually wanting to speed up the heartbeat and this makes that easy.

I restricted heeartbeats to >=1s per our tiny chat. That restriction can get slower to... 10s? I've tagged this for the dev call to discuss with others

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 3 above)
Comment (WickedShell) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Satellite needs you to be able to persistently control the messaging rates, or you are going to burn a lot of money on turning stuff off on boot....

**Comment by magicrub** [comment]
per the dev call, we decided to treat a heartbeat like any other msg but force it to 1Hz at boot. That means you can set it to zero or super slow if desired.

tested values -1, 0, 200, 1000, 5000 all work as expected.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 3 above)
Comment (magicrub) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: not supported, for now a reset will set to 1Hz. If we want the value to be persistent, all values should be set to that and that's work for a new PR.

**Comment by peterbarker** [comment]
I've just realised what the problem with this was.  Sadly, this did cross my mind before merge, but I forgot again.

So the issue is that after this patch heartbeats are subject to slowdown on bad links.  Previously they were exempt.  They were exempt because they are important for the protocol to work.

I don't think we need to do anything about this right now - but we do need to watch out for incidents this might explain.

**Pull request merged**